drivers/rtc/rtc_pcf8523.c

```diff
@@ -0,0 +1,962 @@
+				   BIT(config->int1.pin));
+
+		err = gpio_add_callback_dt(&config->int1, &data->int1_callback);
+		if (err != 0) {
+			LOG_ERR("failed to add GPIO callback (err %d)", err);
+			return -ENODEV;
+		}
+
+		tid = k_thread_create(&data->int1_thread, data->int1_stack,
+				      K_THREAD_STACK_SIZEOF(data->int1_stack),
+				      (k_thread_entry_t)pcf8523_int1_thread, (void *)dev, NULL,
+				      NULL, CONFIG_RTC_PCF8523_THREAD_PRIO, 0, K_NO_WAIT);
+		k_thread_name_set(tid, "pcf8523");
+
+		/*
+		 * Defer GPIO interrupt configuration due to INT1/CLKOUT pin sharing. This allows
+		 * using the CLKOUT square-wave signal for RTC calibration when no alarm/update
+		 * callbacks are enabled (and not configured as a wakeup-source).
+		 */
+	}
+#endif /* PCF8523_INT1_GPIOS_IN_USE */
+
+	/*
+	 * Manually initialize the required PCF8523 registers as performing a software reset will
+	 * reset the time circuits.
+	 */
+
+	/* Read registers PCF8523_CONTROL_1 through PCF8523_CONTROL_3 */
+	err = pcf8523_read_regs(dev, PCF8523_CONTROL_1, &regs, sizeof(regs));
+	if (err != 0) {
+		return -ENODEV;
+	}
+
+	/* Set quartz crystal load capacitance */
+	if (config->cap_sel) {
+		regs[0] |= PCF8523_CONTROL_1_CAP_SEL;
+	} else {
+		regs[0] &= ~(PCF8523_CONTROL_1_CAP_SEL);
+	}
+
+	/* Use 24h time format */
+	regs[0] &= ~(PCF8523_CONTROL_1_12_24);
+
+	/* Disable second, alarm, and correction interrupts */
+	regs[0] &= ~(PCF8523_CONTROL_1_SIE | PCF8523_CONTROL_1_AIE | PCF8523_CONTROL_1_CIE);
+
+	if (config->wakeup_source) {
+		/*
+		 * Always set AIE if wakeup-source. This allows the RTC to wake up the system even
+		 * if the INT1 interrupt output is not directly connected to a GPIO (i.e. if
+		 * connected to a PMIC input).
+		 */
+		regs[0] |= PCF8523_CONTROL_1_AIE;
+	}
+
+	/* Clear interrupt flags (except alarm flag, as a wake-up alarm may be pending) */
+	regs[1] &= ~(PCF8523_CONTROL_2_CTAF | PCF8523_CONTROL_2_CTBF | PCF8523_CONTROL_2_SF);
+
+	/* Disable watchdog, timer A, and timer B interrupts */
+	regs[1] &= ~(PCF8523_CONTROL_2_WTAIE | PCF8523_CONTROL_2_CTAIE | PCF8523_CONTROL_2_CTBIE);
+
+	/* Configure battery switch-over function */
+	regs[2] &= ~(PCF8523_CONTROL_3_PM_MASK);
+	regs[2] |= FIELD_PREP(PCF8523_CONTROL_3_PM_MASK, config->pm);
+
+	/* Clear battery status interrupt flag */
+	regs[2] &= ~(PCF8523_CONTROL_3_BSF);
+
+	/* Disable battery status interrupts */
+	regs[2] &= ~(PCF8523_CONTROL_3_BSIE | PCF8523_CONTROL_3_BLIE);
+
+	/* Write registers PCF8523_CONTROL_1 through PCF8523_CONTROL_3 */
+	err = pcf8523_write_regs(dev, PCF8523_CONTROL_1, &regs, sizeof(regs));
+	if (err != 0) {
+		return -ENODEV;
+	}
+
+	/* Disable watchdog and countdown timers, configure IRQ level*/
+	tmr_clkout_ctrl = 0U;
+
+	if (config->wakeup_source) {
+		/* Disable CLKOUT */
+		tmr_clkout_ctrl |= PCF8523_TMR_CLKOUT_CTRL_COF_MASK;
+	} else {
+		/* Configure CLKOUT frequency */
+		tmr_clkout_ctrl |= FIELD_PREP(PCF8523_TMR_CLKOUT_CTRL_COF_MASK, config->cof);
```

Comment:
It should be possible to do the FIELD_PREP at compile time and store the already shifted value into `config->cof`.